### PR TITLE
Reducing should work (theoretically)

### PR DESCRIPTION
**Pull request by shrutiyer**
I have set everything up for 

* reducing to start adding as soon as some mapper calls it
* master to keep track of how many mapper tasks still need to be finished
* terminate the reducer factory after n millisec
* right before termination, reducer gives the final count to master

What needs to be done:

* Write the word count to file. I already wrote a function to do that. We just need to call it from somewhere once the last reducer has reported the count back. And we should be done :fireworks: :champagne: 

Note: My IntelliJ was weirding out. It won't compile my code :frowning_face:

### CODE REVIEW

File: src/Master.java [hunk 1]
```diff
@@ -54,16 +57,36 @@ public Master(String myIp, String path, List<String> IPs) throws RemoteException
...
-    // }
+    @Override
+    public void markMapperDone() throws RemoteException {
+        // keeps track of how many mappers need to be still processed
+        this.mapTaskIndex--;
+        if (this.mapTaskIndex <= 0) {
+            long now = System.currentTimeMillis();
+            long expectedElapsedTime = now + 5000; // Wait for 5 secs
+            // TODO: Change how this is happening
+            while (now < expectedElapsedTime) {
```
Comment (davidabrahams) [inline src/Master.java]: Switch this to a sleep

File: src/Master.java [hunk 2]
```diff
@@ -75,7 +98,7 @@ public void startWordCount() throws IOException, NotBoundException, AlreadyBound
             Registry reg = LocateRegistry.getRegistry(IPList.get(i));
             iMapper factory = (iMapper) reg.lookup("map_manager");
             factory.createMapTask("map_task_" + j).processInput(line, this);
-            j++; i++;
+            j++; i++; mapTaskIndex++;
```
Comment (davidabrahams) [inline src/Master.java]: todo: set a boolean that is if we're still assigning tasks, use it to terminate mappers

File: src/Master.java [hunk 3]
```diff
@@ -54,16 +57,36 @@ public Master(String myIp, String path, List<String> IPs) throws RemoteException
...
+        // keeps track of how many mappers need to be still processed
+        this.mapTaskIndex--;
+        if (this.mapTaskIndex <= 0) {
+            long now = System.currentTimeMillis();
+            long expectedElapsedTime = now + 5000; // Wait for 5 secs
+            // TODO: Change how this is happening
+            while (now < expectedElapsedTime) {
+                now = System.currentTimeMillis();
+            }
+            iReducer factory = (iReducer) reg.lookup("reduce_manager");
```
Comment (davidabrahams) [inline src/Master.java]: reg is not correct variable here

File: src/Master.java [hunk 4]
```diff
@@ -54,16 +57,36 @@ public Master(String myIp, String path, List<String> IPs) throws RemoteException
...
+            factory.terminate();
+        }
+    }
```
Comment (davidabrahams) [inline src/Master.java]: todo: reduce tasks are not being killed yet, only managers

File: src/Master.java [hunk 5]
```diff
@@ -54,16 +57,36 @@ public Master(String myIp, String path, List<String> IPs) throws RemoteException
...
 
     @Override
     public void receiveOutput(String key, int value) throws RemoteException {
         // reducers call this function when they are done counting
+        this.wordCountMap.put(key, value);
+        this.writeWordCountToFile();
     }
 
+    private void writeWordCountToFile() {
```
Comment (davidabrahams) [inline src/Master.java]: todo: actually call this

**Review by davidabrahams — CHANGES_REQUESTED** [review]

File: src/Master.java (hunk 1 above)
Comment (shrutiyer) [reply src/Master.java]: Done

File: src/Master.java (hunk 3 above)
Comment (shrutiyer) [reply src/Master.java]: Done

File: src/Master.java (hunk 4 above)
Comment (shrutiyer) [reply src/Master.java]: Done

**Pull request merged**